vsphere/resource_vsphere_host.go

```diff
@@ -0,0 +1,683 @@
+	if err != nil {
+		return err
+	}
+	d.Set("maintenance", maintenanceState)
+
+	// Retrieve host's properties.
+	log.Printf("[DEBUG] Got host %#v", hs)
+	host, err := hostsystem.Properties(hs)
+	if err != nil {
+		log.Printf("[DEBUG] Error while retrieving properties for host %s. Error: %s", hostID, err)
+		return err
+	}
+
+	if host.Parent != nil {
+		d.Set("cluster", host.Parent.Value)
+	} else {
+		d.Set("cluster", "")
+	}
+
+	connectionState, err := hostsystem.GetConnectionState(hs)
+	if err != nil {
+		return err
+	}
+
+	if connectionState == types.HostSystemConnectionStateDisconnected {
+		// Config and LicenseManager cannot be used while the host is
+		// disconnected.
+		d.Set("connected", false)
+		return nil
+	}
+	d.Set("connected", true)
+
+	lockdownMode, err := hostLockdownString(host.Config.LockdownMode)
+	if err != nil {
+		return nil
+	}
+
+	log.Printf("Setting lockdown to %s", lockdownMode)
+	d.Set("lockdown", lockdownMode)
+
+	lm := license.NewManager(client.Client)
+	am, err := lm.AssignmentManager(context.TODO())
+	if err != nil {
+		return err
+	}
+	licenses, err := am.QueryAssigned(context.TODO(), hostID)
+	if err != nil {
+		return err
+	}
+
+	licenseKey := d.Get("license").(string)
+	licFound := false
+	for _, lic := range licenses {
+		if licenseKey == lic.AssignedLicense.LicenseKey {
+			licFound = true
+			break
+		}
+	}
+
+	if !licFound {
+		d.Set("license", "")
+	}
+
+	return nil
+}
+
+func resourceVsphereHostCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*VSphereClient).vimClient
+
+	clusterID := d.Get("cluster").(string)
+	ccr, err := clustercomputeresource.FromID(client, clusterID)
+	if err != nil {
+		log.Printf("[DEBUG] Error while searching cluster %s. Error: %s", clusterID, err)
+		return err
+	}
+
+	hcs := types.HostConnectSpec{
+		HostName:      d.Get("hostname").(string),
+		UserName:      d.Get("username").(string),
+		Password:      d.Get("password").(string),
+		SslThumbprint: d.Get("thumbprint").(string),
+		Force:         d.Get("force").(bool),
+	}
+
+	licenseKey := d.Get("license").(string)
+
+	lm := license.NewManager(client.Client)
+	ll, err := lm.List(context.TODO())
+	if err != nil {
+		return err
+	}
+
+	licFound := false
+	for _, l := range ll {
+		if l.LicenseKey == licenseKey {
+			licFound = true
+			break
+		}
+	}
+	if !licFound {
+		return fmt.Errorf("license key supplied (%s) did not match against known license keys", licenseKey)
+	}
+
+	var connectedState bool
+	val := d.Get("connected")
+	if val == nil {
+		connectedState = true
+	} else {
+
+		connectedState = val.(bool)
+	}
+
+	task, err := ccr.AddHost(context.TODO(), hcs, connectedState, &licenseKey, nil)
+	if err != nil {
+		log.Printf("[DEBUG] Error while adding host with hostname %s to cluster %s.  Error: %s", d.Get("hostname").(string), clusterID, err)
+	}
+	task.Wait(context.TODO())
+
+	var to mo.Task
+	err = task.Properties(context.TODO(), task.Reference(), nil, &to)
```

Comment:
Same question as above on non-success state that doesn't result in an error.

Reply:
See above :)